User-defined constrained parameterizations (#967)
stan-dev/docs: src/stan-users-guide/user-defined-constraints.qmd

Reviewer: It might make sense to reassess https://mc-stan.org/docs/stan-users-guide/reparameterization.html#changes-of-variables and see if that should just be blended into this new chapter (or at least cross-referenced)

---
pagetitle: User-Defined Constrained Parameterizations
---

# User-Defined Constrained Parameters

This chapter introduces techniques for defining model
parameterizations that satisfy constraints not handled by Stan's
built-in types.  There are several general strategies used to do this.
The most straightforward approach is to build the type out of Stan's
existing constrained types and transforms, as that will let Stan
automatically work out the change-of-variables.  The more general
approach is to code a user-defined transform function and either

> adjust for then change-of-variables or not, depending on the intended

Reviewer 2: ```suggestion
adjust for the change-of-variables or not, depending on the intended
```

distribution.

## Stan's approach to constrained parameters {#sec-stan-constraints}

Stan's built-in types are designed so that the default distribution on
a parameter is uniform over parameter values satisfying the declared
constraint.  For example, variables of type `real<lower=-1, upper=1>`
are implicitly assigned a uniform prior over their defined range, here
$(-1, 1)$.  For unbounded types, such as `real<lower=0>`, the implicit
prior is uniform, but improper (i.e., the integral over its support
diverges).

> The way this is done is by applying a change-of-variables correction.

Reviewer 2: I would avoid the use of `this`.

```suggestion
Enforcing a uniform implicit prior on constrained parameters requires applying a change-of-variables correction.
```

For example, consider a positive-constrained variable $\theta \in (0,
\infty)$.  This variable is first log transformed to $\phi = \log
\theta \in (-\infty, \infty)$; there are alternative smooth, monotonic
bijections from $(0, \infty)$ to $(-\infty, \infty)$, but Stan uses
the natural logarithm.

The underlying sampler works with the unconstrained variable $\phi$

> that results from transforming the constrained parameter..

Reviewer 2: ```suggestion
that results from transforming the constrained parameter.
```

> In order for the distribution of $\theta$ to be uniform over $(0,

Reviewer 2: ```suggestion
In order for the distribution $p(\cdot)$ of $\theta$ to be uniform over $(0,
```

\infty)$, a change-of-variables correction is needed.  The general
form for univariate variables $\theta$ transformed to $\phi =
f(\theta)$ is
$$
p(\phi)
= p(\theta)
\cdot \left| \frac{\textrm{d}}{{\textrm d}\phi} \theta
\right|,

> $$

Reviewer 2: Should you define `p` before this equation?

> $$

Reviewer 2: I don't love the substitution of $\phi$ for $f\left(\theta\right)$ imo I think this reads nicer as 

$$
p\left(f\left(\theta\right)\right)
= p(\theta)
\cdot \left| \frac{\textrm{d f}}{{\textrm d}\theta} \theta
\right|,
$$

Though applying that through the line right after this I'm not totally understanding the equation. And what is $p_\Phi$?

$$
\begin{align}
p_\Phi(\log\left(\theta\right)) &= p_\Theta\left(\exp\left(\log\left(\theta\right)\right)\right) \cdot \left| \frac{\textrm{d}}{\textrm{d} \log\left(\theta\right)} \exp\left(\log\left(\theta\right)\right) \right|
\\
&= p_\Theta(\exp\left(\log\left(\theta\right)\right)) \cdot \exp\left(\log\left(\theta\right)\right)
\end{align}
$$

Wouldn't the last line reduce $\exp(\log(\cdot))$ so that the final line just becomes 

$$
p_\Theta(\theta) \cdot \theta
$$

Maybe I'm just not understanding something?

where $\theta = f^{-1}(\phi)$.  Applying this to the case at hand,
where $\theta \in (0, \infty)$ and $f(\theta) = \log \theta$,
yields $\theta = f^{-1}(\phi) = \exp(\phi)$, so that
\begin{align}
p_\Phi(\phi) &= p_\Theta(\exp(\phi)) \cdot \left| \frac{\textrm{d}}{\textrm{d} \phi} \exp(\phi) \right|
\\[4pt]

> &= p_\Theta(\exp(\phi)) \cdot \exp(\phi).

Reviewer 2: ```suggestion
&= p_\Theta(\exp(\phi)) \cdot \exp(\phi)
```

in the model block, as the example worked in the next section
demonstrates.

### When is a change-of-variables adjustment required?

A change-of-variables adjustment _must be applied_ when the
constrained parameters (i.e., the result of the inverse transform) are
required to have a uniform baseline distribution.  For example, this
will be the case if a constrained parameter is given a distribution
statement in the model block, as is done for `sigma` in the following
example.

```stan
parameters {
  real<lower=0> sigma;
}
model {
  sigma ~ lognormal(0, 1);
}
```

In this case, the density produced for $\log \sigma$ by Stan will be
$$
\log p(\log \sigma)
= \log \textrm{lognormal}(\exp(\log \sigma) \mid 0, 1) + \log \sigma.
$$
The result is that $\sigma = \exp(\log \sigma)$ has a
$\textrm{lognormal}(0, 1)$ distribution.

The change-of-variables can be coded directly, as shown in the following
Stan program, which lays out the steps explicitly that Stan performs
implicitly in the original program where `sigma` was declared as a parameter
of type `real<lower=0>`.

```stan
parameters {
  real log_sigma;
}
transformed parameters {

>   real<lower=0> sigma = exp(log_sigma);

Reviewer: It might be worth explicitly (re)-stating that the `lower=` syntax only performs checks in blocks other than parameters

is placed on the underlying parameterization.  An alternative way to
produce a lognormally distributed variable `sigma` in a Stan
program is as follows.

```stan
parameters {
  real log_sigma;
}
transformed parameters {
  real<lower=0> sigma = exp(log_sigma);
}
model {
  log_sigma ~ normal(0, 1);
}
```

This model places exactly the same distribution on `sigma` as the two
Stan programs in the previous section did.  Tracing through the math,
it can be seen that the lognormal distribution applies the change of
variables itself.  Applying the `jacobian += log_sigma`
change-of-variables adjustment in this program would be wrong in that
it would no longer produce a `sigma` with a lognormal distribution.

The other situation in which a Jacobian adjustment can be dropped is
when the transform is a linear function of parameters for constant
offsets and multipliers.  For example, if $\alpha = m + s \cdot \beta$
and $m$ and $s$ are constants, then the change of variables adjustment
will be the constant $\log s$, which can be dropped.  If $s$ were a
parameter, then the change-of-variables adjustment would no longer
be constant and thus should be included.

> ## Specialized matrices

Reviewer: I think that this title is probably too broad (i.e. does it imply that e.g. orthogonal matrices are not specialized, since they're in a different section?)

The unifying theme seems to be that these are just simple reshapes or linear functions of vectors, as opposed to the later matrices in the chapter which are not. Maybe the title should therefore be "Simple matrix variants"?

There are dozens if not hundreds of specialized matrix forms in use,
some of the most popular of which will be described in this section.

> 

Reviewer 2: Kind of a 100 G-force of a transition from change of variables to bespoke matrices

tridiagonal matrix has $3 \cdot N - 2$ free entries and the rest are
structural zeros.  The matrix is filled from the vector in row
major order.

```stan
functions {
  matrix tridiagonal_matrix(vector x, int N) {
    if (rows(x) != 3 * N - 2) {
      reject("rows(x) = ", rows(x),
             "; but 3 * N - 2 = ", 3 * N - 2);
    }
    if (N == 1) return [[x[1]]];
    matrix[N, N] y;
    y[1, 1] = x[1];
    y[1, 2] = x[2];
    int next = 3;
    for (n in 2:(N - 1)) {
      y[n, n - 1] = x[next];
      y[n, n] = x[next + 1];
      y[n, n + 1] = x[next + 2];
      next += 3;
    }
    y[N, N - 1] = x[next];
    y[N, N] = x[next + 1];
    return y;
  }
}
parameters {
  vector[3 * N - 2] alpha;
}
transformed parameters {
  matrix[N, N] tri_alpha = tridiagonal_matrix(alpha, N);
}  
```

### Toeplitz matrices

A Toeplitz matrix has the same value along the diagonal.  Thus
an $M \times N$ Toeplitz matrix has $M + N - 1$ free variables.  These
are filled in from the lower left corner $(M, 1)$ to the upper right $(1, N)$.

```stan
functions {
  matrix toeplitz_matrix(vector x, int M, int N) {
    if (num_elements(x) != M + N - 1) {

>       reject("x must have M + N - 1 elements",

Reviewer: I think these `reject`s should be `fatal_error`s, since the sizes will never change there is no recovery possible

    return y;
  }
}
parameters {
  vector[M + N - 1] alpha;
}
transformed parameters {
  matrix[M, N] toeplitz_alpha = toeplitz_matrix(alpha);
}
```

To verify that the lower left corner is assigned `x[1]`, note that the
lower left corner has `i` equal to M and `j` equal to 1, so `M + j -
i` evaluates to 1, as intended. To verify the upper right is assigned
`x[M + N - 1]`, note that the upper upper right has `i` equal to 1 and
`j` equal to `N`, so `M + j - i` is `M + N - 1`. To verify the
diagonals are aligned, note that the requirement is that `M[i, j]` is
the same as `M[i - 1, j + 1]`, which it is because `M + j - i` is
equal to `M + (j - 1) - (i - 1)`.

### Lower-triangular matrices

An $N \times N$ matrix $A$ is lower triangular if $A_{i, j} = 0$ for
all $j > i$. There are $N + \binom{N}{2}$ variables required to
construct a lower-triangular matrix. Because Stan cannot convert a
real value to an integer, the dimension of the resulting matrix is
required as an argument.

```stan
functions {
  matrix lower_triangular_matrix(vector x, int N) {
    if (cols(x) != N + binom(N, 2)) {
      reject("cols(x) = ", cols(x),
             "; but N + binom(N, 2) = ", N + binom(N, 2));

>     }

Reviewer: `binom` is not a function; it's `choose` (which is correctly used in the parameters block below)

    matrix[N, N] y = rep_matrix(0, N, N);  // fills zero entries
    int next = 0;
    for (i in 1:N) {
      for (j in 1:i) {
        y[i, j] = x[next];
        next += 1;
      }
    }    
    return y;
  }
}
parameters {
  vector[choose(N, 2) + N] alpha;
}
transformed parameters {
  matrix[N, N] lower_tri_alpha = lower_triangular_matrix(alpha, N);
}
```

Upper-triangular matrices could be constructed by transposing lower-triangular
matrices, but it's more efficient to code them directly with a modified
form of the `lower_triangular_matrix` function.

### Symmetric matrices

An $N \times N$ matrix $A$ is symmetric if $A = A^\top$ (i.e.,
$A_{i,j} = A_{j, i}$ for all $1 \leq i, j \leq N$). As with a lower
triangular matrix, there are $N + \binom{N}{2}$ free variables in a
symmetric matrix.  To build a symmetric matrix, it suffices to
build a lower-triangular matrix, then copy the entries below the
diagonal into their transposed position above the diagonal.

```stan
functions { 
  matrix symmetric_matrix(vector x, int N) {

>     matrix[N, N] y = lower_triangular_matrix(x, N);

Reviewer: ```suggestion
    matrix[N, N] y = lower_triangular_matrix(x, N); // defined as above
```

Or any other indication that this is not a built-in and therefore this example is not fully self-contained

Building a skew symmetric matrix is like building a symmetric matrix only the
above-diagonal entries are negated.

```stan
functions {
  matrix skew_symmetric_matrix(vector x, int N) {
    matrix[N, N] y = lower_triangular_matrix(x, N);
    for (i in 1:N) {
      for (j in (i + 1):N) {
        y[j, i] = -y[i, j];
      }
    }    
    return y;
  }
}
parameters {
  vector[choose(N, 2) + N] alpha;
}
transformed parameters {
  matrix[N, N] symmetric_alpha = skew_symmetric_matrix(alpha, N);
}
```

## Ordered simplexes

Suppose a prior on simplexes states that the probabilities in the
simplex are in ascending order.  Stan provides an ordered vector type
and a simplex vector type, but it does not provide an ordered simplex
type.  Stan does provide built-in functions that carry out the
transforms and change-of-variables adjustments, though. The simplest
way to define a uniform distribution over ordered simplexes is as
follows.

```stan
data {
  int<lower=1> N;
}
parameters {
  positive_ordered[N - 1] theta_raw;
}
transformed parameters {
  simplex[N] theta = simplex_jacobian(-theta_raw);
}
```

There are only $N-1$ underlying parameters to produce a simplex

> variable $\theta$.  The `pos_ordered` constraint produces a sequence

Reviewer: ```suggestion
variable $\theta$.  The `positive_ordered` constraint produces a sequence
```

of positive values `theta_raw[1]`, `theta_raw[2]`, to `theta_raw[N -
1]` in increasing order.  This means `-theta_raw[1]`, `-theta_raw[2]`,
to `-theta_raw[N]` will be in decreasing order.  The built-in simplex
transform will reverse that order again to produce simplexes in
increasing order.  In the program above, `simplex[N]` could have been
declared `positive_ordered[N]` without breaking anything---the type on

> transformed parameters only does error checking.

Reviewer: Ah, so you do mention it here -- probably worth pulling up earlier (or repeating)

An alternative way to program this model would be to use raw parameters and
do both transforms manually.

```stan
data {
  int<lower=1> N;
}
parameters {
  vector[N - 1] theta_raw;
}
transformed parameters {
  simplex[N] theta

>       = simplex_jacobian(positive_ordered_jacobian(theta_raw));

Reviewer: Is this missing a negative compared to the above?

}
```

Enforcing an increasing sequence pattern on the simplex leads to a
uniform distribution with small means on the early components and
relatively large means on the later ones.  For example, with `N` set
to 2, the posterior means are $[0.25 \ 0.75]$, with `N` set to 5, they
are $[0.05 \ 0.08 \ 0.12 \ 0.23 \ 0.52]$, and set to 10, they are
$[0.02 \ 0.02 \ 0.03 \ 0.04 \ 0.05 \ 0.07 \ 0.10 \ 0.14 \ 0.20 \

> 0.33]$.  To derive a more uniform distribution, they can be given

Reviewer: I find this paragraph confusing -- if they have a uniform distribution as stated in the first sentence, how can it be 'more uniform'?

a Dirichlet prior that's skewed in the opposite direction.  For example,
consider adding the following model block for the case where `N` is 5.
```stan
model {
  theta ~ dirichlet([16, 8, 4, 2, 1]');
}
```
The resulting posterior means are now $[0.17 \ 0.18 \ 0.19 \ 0.21 \
0.25]$.

It feels counterintuitive that giving such a skewed Dirichlet prior
leads to an almost uniform ordered distribution, but that's just the
natural result of two densities pulling in opposite directions.  The
uniform distribution over ordered simplexes leads to heavily skewed
means.  The explicit Dirichlet prior in the `model` block then pulls
it back closer to uniformity.  This overall point is characteristic of
constrained distributions, which will not behave like uniform
distributions on the unconstrained space when the transforms are
non-linear.

## Correlation matrices with structural zeros

Stan does _not_ provide a built-in type for correlation matrices that

> have structural zeros (i.e., correlations known to be zero). Such

Reviewer: This reads awkwardly to me, since the whole chapter is things Stan does not provide. Maybe "While Stan does provide built-in types for correlation matrices, there is not support for structural zeros..."

multiplied out have a uniform distribution over correlation matrices.
The `lkj_corr_cholesky` distribution builds in the
change-of-variables adjustment resulting from transforming `L1` to
`Omega1`, then applies the LKJ distribution to `Omega1` (conceptually;
computationally it's more efficient than that).

With Stan's built-in constraint transformation functions, another way
of specifying a parameter to be the Cholesky factor of a correlation
matrix is as follows.

```stan
parameters {
 vector[choose(5, 2)] L3_raw;
}
transformed parameters {
  cholesky_factor_corr[5] L3
      = cholesky_factor_corr_jacobian(L3_raw, 5);
}
```

In this program, `L3` will have a uniform distribution over Cholesky
factors of correlation matrices, matching `L1` in the original
program.  In particular, it can be given an LKJ-correlation distribution
to make the resulting correlation matrices LKJ distributed.

### Specifying structural zeros

Stan does not support user-defined types. The closest approximation is
user-defined `_jacobian` transform functions that increment the
Jacobian accumulator with a change-of-variables adjustment. The rest

> of this section shows how to use this mechanism to develop correlation

Reviewer: This also feels awkward -- I'm guessing what happened is that this section was written first and the rest of the chapter grew around it, but now it feels out of place since this kind of introductory material was covered earlier

Jacobian adjustment to render the resulting distribution uniform on
Cholesky factors of correlation matrices with the required structural zeros.

The transform function relies on a simple helper function that takes
the zeros indicator, a row index, and a value and returns an array of
columns up to the row index minus one that match the value.  For
example, 

```stan
  /**
   * Return an array of column indexes from the specified row
   * of the zero indicator matrix that have the specified value.
   *
   * @param i The row index.
   * @param is_zero The zero indicator matrix.
   * @param val Indicator value to match (0 or 1).
   * @return Ascending array of indexes of columns that match.
   */
  array[] int columns_matching(int i, array[,] int is_zero,
                               int val) {
    int n = 0;
    for (c in 1:(i - 1)) {
      n += (is_zero[i, c] == val);
    }
    array[n] int cols_match;
    int pos = 0;
    for (c in 1:(i - 1)) {
      if (is_zero[i, c] == val) {
        pos += 1;
        cols_match[pos] = c;
      }
    }
    return cols_match;
  }
```
The documentation follows the style of the multi-language
documentation generator

> [Doxygen](https://en.wikipedia.org/wiki/Doxygen).

Reviewer: This feels unnecessary to call out, I think it's readable even to someone unfamiliar

The next helper function extracts the column indexes that match a
given value for a given row of the zero correlation indicator array.

```stan
/**
 * Return the columns `c` in `1:(i - 1)` for which `row_ind[c] ==
 * val`, in ascending order.  Used to split the columns of row `i`
 * of a Cholesky factor into structurally zero columns (`val == 1`)
 * and free columns (`val == 0`).
 *
 * @param i Row index.
 * @param row_ind Indicator array of size >= i - 1 (row i of the
 *   zero-indicator matrix).
 * @param val Indicator value to match (0 or 1).
 * @return Ascending array of matching column indices.
 */

> array[] int columns_matching(int i, array[] int row_ind, int val) {

Reviewer: It looks like only the first of these helpers is actually called?

  int n = 0;
  for (c in 1:(i - 1)) {
    n += (row_ind[c] == val);
  }
  array[n] int idx;
  int pos = 0;
  for (c in 1:(i - 1)) {
    if (row_ind[c] == val) {
      pos += 1;
      idx[pos] = c;
    }
  }
  return idx;
}
```

The next function does all the heavy lifting.  It has a long documentation
string explaining what it does.

```stan
  /**
   * Transform unconstrained parameters to the Cholesky factor of a
   * correlation matrix Omega = L * L' with the specified structural
   * zeros.
   * 
   * A change-of-variables adjustment is applied to the log Jacobian
   * accumulator so that if `raw` has an (improper) uniform
   * distribution, the free entries of `L` (the
   * strictly-lower-triangular entries not forced by the structural
   * zeros) are uniformly distributed over their feasible region.
   *
   * @param D The number of dimensions.
   * @param raw The unconstrained parameters (size choose(D, 2)).
   * @param is_zero A D x D indicator array with `is_zero[i, j]`
   *    indicating zero correlation between variables `i` and `j`.
   * @return The Cholesky factor of a correlation matrix satisfying
   *   the structural zeros.
   */

> ```
> 
> The transform code itself follows.
> 
> ```stan

Reviewer: ```suggestion
```

 * specified concentration for the specified Cholesky factor
 * of a correlation matrix, corrected for the specified
 * structural zero counts.
 *
 * @param L Cholesky factor of a correlation matrix.
 * @param eta LKJ concentration (eta > 0).
 * @param col_zeros Per-column counts of below-diagonal
 *     structural zeros.
 * @return Log density of `L`, corrected for structural zeros.
 */
real lkj_corr_cholesky_zeros_lpdf(matrix L, real eta,
                                  array[] int col_zeros) {
  return lkj_corr_cholesky_lpdf(L | eta)
    - to_row_vector(col_zeros) * log(diagonal(L));
}
```

### Stan program using the transform

With all the heavy lifting done in functions, the Stan program that depends
on them is simple.

```stan
data {
  int<lower=2> D;                            // dimensions
  real<lower=0> eta;                         // LKJ concentration
  array[D, D] int<lower=0, upper=1> zeros;   // lower tri zeros
}
transformed data {
  array[D] int<lower=0> col_zeros;
  for (j in 1:D) {
    col_zeros[j] = sum(zeros[(j + 1):D, j]);  // below diag only
  }
  int N_zero = sum(col_zeros);
}
parameters {
  vector[choose(D, 2) - N_zero] raw;
}
transformed parameters {

>   matrix[D, D] L_Omega

Reviewer: The style so far has been to use the constrained type name in transformed parameters to get the extra checking that results, which is worth continuing IMO

distribution.  

For each pair of non-zero indexes below the diagonal, there is a
Jacobian adjustment for the change of variables involved in the
constraining (inverse) transform due to the lower/upper bounding. This
adjustment is handled implicitly by Stan's built-in lower/upper-bound
transform.

This defines a variable `L_Omega` that is uniform over Cholesky
factors of correlation matrices with the specified structural zeros.
The final result is an `Omega` with an $\textrm{LKJ}(\nu)$
distribution restricted to correlation matrices with the specified
structural zeros. This works no matter what $\eta > 0$ is (e.g.,
setting `eta` to 1 produces a uniform distribution).

As part of a larger application, the correlation matrix or its
Cholesky factor are typically paired with a vector of scale parameters
to produce a covariance matrix with the specified pattern of zeros.

As an example, consider the zeros matrix specified earlier and $\eta = 1$.
Here's a representative
draw,
$$
\Omega^{(102)}
=
\begin{bmatrix}
1 & 0 & -0.402 & 0.395 & 0 \\
0 & 1 & 0 & -0.708 & 0.590 \\
-0.402 & 0 & 1 & -0.185 & 0.296 \\
0.395 & -0.708 & -0.185 & 1 & 0 \\
0 & 0.590 & 0.296 & 0 & 1
\end{bmatrix}
$$

> The 0 values are truly zero, which is remarkable for numerical solves
> involving double-precision floating point; the corresponding summary
> output has zero standard deviation across the draws.  For example,
> here's one line.
> ```sh
>                       Mean      MCSE        StdDev 
> Omega[1,2]    0.000000e+00       NaN  0.000000e+00 
> ```
> 

Reviewer: ```suggestion

```

Feels unnecessary to me

## Orthogonal, rotation, and reflection matrices

Stan has a built-in `unit_vector` type that produces vectors where the
sum of the squared entry is 1.  Those points live on a unit
hypersphere.  This section generalizes this notation to orthonormal
matrices, rotation matrices, and reflection matrices.

### Orthogonality and orthonormality

A pair of $N$-vectors $u, v$ is _orthogonal_ if $u^\top \cdot v = 0$,
which entails a cosine of zero.  Orthogonal is the $N$-dimensional
generalization of the opposite and adjacent edges of a right triangle.
An $N$-vector $u$ has Euclidean _length_ $||u||_2 = \sqrt{u^\top \cdot
u}$.  Thus it has unit length if $||u||_2 = 1$.  A pair of vectors is
_orthonormal_ if they are orthogonal and unit length.

An _orthonormal matrix_ is an $M \times N$ matrix in which the columns

> are orthonormal (waring: in some treatments rows are orthonormal).  In

Reviewer: ```suggestion
are orthonormal (warning: in some treatments rows are orthonormal).  In
```

I realize you get into the square case below, but it might be worth calling out that these definitions would be equivalent then

> a somewhat confusing terminological convention, if $M = N$, the matrix

Reviewer: Some sources use these terms interchangeably. Notably wikipedia seems to treat the terms as equivalent and _both_ only applying to square matrices.

I have heard 'left orthogonal matrix' for a matrix whose columns form an orthogonal set (the definition you're giving here for orthonormal matrix), which is nice and unambiguous

is further said to be _orthogonal_ (the confusion arises because the
columns are orthogonal in an orthonormal matrix even if $M \neq N$).

Stan's built-in unit vector type (`unit_vector`) produces the boundary
case of an orthonormal matrix with only a single column.

### Rotation and reflection

If $y$ is a point on a hypersphere (e.g., a unit hypersphere) and $X$
is orthogonal, then $y \cdot X$ is on the same hypersphere.  A point
on the hypersphere can be derived by either rotation or reflection.

An orthogonal matrix $X$ with $\det X = 1$ is a _rotation matrix_ in
that its action on a vector $y$, $y \cdot X$, is a rotation.  You
can get from any point on the sphere to any other point on a sphere
through a rotation.

An orthogonal $X$ with $\det X = -1$ is said to be a _reflection
matrix_.  A reflection matrix maps a point through the origin to the
opposite pole.  

### The special orthogonal group

The set of rotation matrices of dimension $N \times N$ forms a group
structure which is written as $\textrm{SO}(N)$ for _special orthogonal
group_.  It's easy to verify that the rotation matrices meet the

> requirements to form a group under multiplication.  First,

Reviewer: Maybe link to a definition for 'group'?

aren't other stochastic initialization bottlenecks.

### Rotations by exponentiating a skew-symmetric matrix

The matrix exponential of an $N \times N$ skew-symmetric matrix is a
member of the special orthogonal group $\textrm{SO}(N)$.  The transform
will be implemented like a built-in Stan transform with a Jacobian
adjustment.  The transform itself is rather opaque, as is the
following helper function.  The helper function computes provides
a numerical stable computation of $2 \cdot
\textrm{sum}\!\left(\log \left| \textrm{sinc}(x / 2)\right|\right)$, where all operations are elementwise. The arithmetic
stability stems from replacing calculations when $x$ is near zero with their
Taylor expansion.

```stan
  real sum_two_log_abs_half_sinc(vector x) {
    int N = rows(x);
    vector[N] h = x / 2;
    vector[N] half_sinc = sin(h) ./ h;
    for (n in 1:N) {
      if (abs(h[n]) < 1e-4) {  // Taylor expansion for x near 0
        real h_sq = square(h[n]);
        half_sinc[n] = 1 - h_sq / 6 + square(h_sq) / 120;
      }
    }
    return 2 * sum(log(abs(half_sinc)));
  }
```  

The transform itself, which takes a vector $v$ of size $\binom{N}{2}$,

> creates a skew-symmetric matrix (the `skew_symmetric` function is defined

Reviewer: ```suggestion
creates a skew-symmetric matrix (the `skew_symmetric_matrix` function is defined
```

in @sec-skew-symmetric-matrices), applies matrix exponential, then adds
the Jacobian adjustment to the Jacobian accumulator.

```stan
/**
* Return the rotation matrix for the specified basis, adding the
* change-of-variables adjustment to the Jacobian accumulator.
*
* @param v Basis vector (size `choose(N, 2)`).
* @param N Dimensions of rotation.
* @return The `N` x `N` rotation matrix for which `v` is the basis
*/
matrix special_orthogonal_jacobian(vector v, int N) {

>   matrix[N, N] S = skew_symmetric(v, N);

Reviewer: ```suggestion
  matrix[N, N] S = skew_symmetric_matrix(v, N);
```

stationary covariance `Gamma`.

The time series model is going to use a simple helper function, which
is not documented here because it follows Stan's naming convention
for reshaping functions.

```stan
array[] row_vector to_array_of_row_vectors(matrix m) {
  int R = rows(m);
  array[R] row_vector[cols(m)] a;
  for (r in 1 : R) {
    a[r] = m[r];
  }
  return a;
}
```  

The model itself is what one would expect given what came before.
The data is the dimensionality, number of time steps, and observations.
The transforms generate a vector of zeros and an identity matrix, as well
as a convenient matrix form of the observations transposed.

```stan
data {
  int<lower=1> D;
  int<lower=2> T;
  array[T] vector[D] y;
}
transformed data {
  vector[D] zeros_D = rep_vector(0, D);
  matrix[D, D] I_D = identity_matrix(D);
  matrix[T - 1, D] y_last;
  for (t in 1 : T - 1) {
    y_last[t] = y[t]';
  }
}
```

The parameters are the raw parameters, which are then transformed
using the functions just defined.

```stan
parameters {
  vector[choose(D, 2) + D * D + D] raw;
}
transformed parameters {
  matrix[D, D] phi;
  matrix[D, D] Gamma;

>   {
>     tuple(matrix[D, D], matrix[D, D]) phi_Gamma
>       = stationary_var1_jacobian(raw, D);
>     phi = phi_Gamma.1;
>     Gamma = phi_Gamma.2;
>   }

Reviewer: We have tuple unpacking

```suggestion
  (phi, Gamma) = stationary_var1_jacobian(raw, D);
```

  matrix[D, D] Sigma = Gamma - quad_form_sym(Gamma, phi);
}
```

The result is an induced uniform distribution over the coefficients
`phi` leading to a stationary VAR(1) process and the stationary
covariance `Gamma`.  These need to be pulled out of the tuple.  This
is one function returning a tuple because there is shared work.  The
extra braces mean that `phi_Gamma` is no longer a top-level
`transformed parameters` variable and thus will not be saved in the
output (it would be redundant).  The model could have been written
directly in terms of `phi_Gamma.1` and `phi_Gamma.2`, but it's easier

> to follow as written.  Note that `phi` is used in its transposed form

Reviewer: Which makes this explanation redundant

here.  All the transposition was to ensure none was needed at run
time, just at data construction time.

```stan
model {
  Gamma ~ inv_wishart(D + 2, I_D);
  to_vector(phi) ~ normal(0, 1);

  y[1] ~ multi_normal(zeros_D, Gamma);
  array[T - 1] row_vector[D] mu
    = to_array_of_row_vectors(y_last * phi);
  y[2: ] ~ multi_normal(mu, Sigma);
}
```

The priors are an inverse Wishart prior on the stationary covariance and a weakly
informative standard normal on the regression coefficients.  These could
have been chosen differently.  To apply an LKJ prior would require factoring out
a correlation matrix from `Gamma` and applying the appropriate change-of-variables
adjustment.  

As a test precaution, the spectral radius (i.e., the maximum absolute eigenvalue)
of `phi` can be computed.  

```stan
generated quantities {
  real<lower=0, upper=1> spectral_radius
    = max(abs(eigenvalues(phi)));
}
```

The constraints set `lower=0` and `upper=1`, because that's the
constraint that leads to stationarity.  There are no additional
constraints on `Gamma` beyond the positive-definiteness required by
its declaration.

## Matrices matching known margins

> Suppose $X$ is an $M \times N$ matrix $X$, the marginals of which are

Reviewer: ```suggestion
Suppose $X$ is an $M \times N$ matrix, the marginals of which are
```
